### PR TITLE
refactor(common/models): LMLayer state management tweak, persistent ModelCompositor

### PR DESCRIPTION
**Pull request by jahorton**
In order to proceed with edit-distance work, one of the first things that we have to do:  allow a persistent `ModelCompositor` that can store intermediate calculations, rather than reconstructing it every time.  

To facilitate this and to allow `ModelCompositor` to be accessible as an LMLayer property when in the correct state, the three states of the LMLayer have now been represented as a TS discriminated union.  The new `.compositor` properly is only accessible on the state object after condition-checking against `state.name == "ready"`.

### CODE REVIEW

File: common/predictive-text/worker/worker-interfaces.ts [hunk 1]
```diff
@@ -118,19 +118,44 @@ interface WordbreakMessage {
...
   handleMessage(payload: IncomingMessage): void;
 }
 
+/**
+ * Represents the pre-model-load state of the LMLayer.
+ */
+interface LMLayerWorkerModellessState {
+  /**
+   * Informative property. Name of the state. Currently, the LMLayerWorker can only
+   * be the following states:
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts:141]:
```suggestion
```

File: common/predictive-text/worker/worker-interfaces.ts [hunk 2]
```diff
@@ -118,19 +118,44 @@ interface WordbreakMessage {
...
+  handleMessage(payload: IncomingMessage): void;
+}
+
+/**
+ * Represents the 'ready' state of the LMLayer.
+ */
+interface LMLayerWorkerReadyState {
+  /**
+   * Informative property. Name of the state. Currently, the LMLayerWorker can only
+   * be the following states:
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts:153]:
```suggestion
```

File: common/predictive-text/worker/worker-interfaces.ts [hunk 3]
```diff
@@ -118,19 +118,44 @@ interface WordbreakMessage {
...
 
 /**
- * Represents a state in the LMLayer.
+ * Represents the unconfigured state of the LMLayer.
  */
-interface LMLayerWorkerState {
+interface LMLayerWorkerUnconfiguredState {
   /**
    * Informative property. Name of the state. Currently, the LMLayerWorker can only
    * be the following states:
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts:129]:
```suggestion
```

File: common/predictive-text/worker/worker-interfaces.ts [hunk 4]
```diff
@@ -118,19 +118,44 @@ interface WordbreakMessage {
   context: Context;
 }
 
+type LMLayerWorkerState = LMLayerWorkerUnconfiguredState | LMLayerWorkerModellessState | LMLayerWorkerReadyState;
```
Comment (eddieantonio) [inline common/predictive-text/worker/worker-interfaces.ts:121]:
```suggestion
/***
 * The LMLayer can be in one of the following states. The LMLayer can only produce predictions in the 'ready' state.
 */
type LMLayerWorkerState = LMLayerWorkerUnconfiguredState | LMLayerWorkerModellessState | LMLayerWorkerReadyState;
```

**Comment by mcdurdin** [comment]
Can I request a different title for the PR? Rather than talking about the implementation, talk about the problem that it is solving.

For example:  "Make ModelCompositor persistent so that it can store intermediate calculations"

**Pull request merged**